[clang-tidy] Fix false positives with deducing this in `readability-convert-member-functions-to-static` check

flovent opened the pull request:
Fixes #141381.

Add check for `DeclRefExpr` which points to an explicit object parameter.

Commit:
[clang-tidy] fix false positives with deducing this in `readability-convert-member-functions-to-static` check

vbvictor commented on clang-tools-extra/test/clang-tidy/checkers/readability/convert-member-functions-to-static-deducing-this.cpp:
```diff
@@ -0,0 +1,14 @@
+  void println(const char *format, const std::string &str) {}
+}
+
+namespace PR141381 {
```
Nit: don't think that this has any profits for readability/maintainability.
AFAIK such namespaces are used for small bug-fixes that are usually placed in big test files, so with such namespaces it's easier to navigate through tests.
Your file is completely new, and there is no need for such namespace.

Commit:
don't traverse all DeclRefEx[r

Commit:
improve test cases based on reviewd feedback

flovent commented on clang-tools-extra/test/clang-tidy/checkers/readability/convert-member-functions-to-static-deducing-this.cpp:
```diff
@@ -0,0 +1,14 @@
+struct Hello {
+  std::string str_;
+
+  void hello(this Hello &self) { std::println("Hello, {0}!", self.str_); }
```
testcases added.

vbvictor commented on clang-tools-extra/clang-tidy/readability/ConvertMemberFunctionsToStatic.cpp:
```diff
@@ -79,10 +79,10 @@ void ConvertMemberFunctionsToStatic::registerMatchers(MatchFinder *Finder) {
-              hasTrivialBody(), isOverloadedOperator(), cxxConstructorDecl(),
-              cxxDestructorDecl(), cxxConversionDecl(), isTemplate(),
-              isDependentContext(),
+              isExplicitObjectMemberFunction(), isExpansionInSystemHeader(),
```
Please place `isExplicitObjectMemberFunction()` after `cxxConversionDecl()`. These predicates are checked from first to last in order, so for efficiency I'd put it after `cxxConversionDecl()`.

vbvictor commented on clang-tools-extra/test/clang-tidy/checkers/readability/convert-member-functions-to-static-deducing-this.cpp:
```diff
@@ -0,0 +1,22 @@
+// RUN: %check_clang_tidy -std=c++23 %s readability-convert-member-functions-to-static %t
```
```suggestion
// RUN: %check_clang_tidy -std=c++23-or-later %s readability-convert-member-functions-to-static %t
```

vbvictor reviewed (approved):
LGTM

Commit:
adjust predicates' order

Commit:
c++23 to c++23-or-later

vbvictor commented:
Could you please rebase on fresh main, and I will land the changes.

flovent commented:
Rebased and changed my email to public.

Different with this PR, I use github noreply email in my latest PR to commit, do I need to rebase to change it or it will automaticly update to my public email when merging like it will transfter to noreply email like [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) says?

carlosgalvezp reviewed (approved):
LGTM!

vbvictor commented:
When merging, your new public email will be used, now it says
"This commit will be authored by flbven@protonmail.com."
